user_guide.adoc

```diff
@@ -1090,7 +1090,117 @@ bytes sparsely.
 
 === Streams and substreams
 
-TODO
+====Introduction and simple example====
+
+A stream is a flow of data from an input file into a parser which is
+generated by a KS script. The parser can request one or more bits of
+data from the stream at a time, but cannot request the same data twice
+and cannot request data be provided out of sequential order. A stream
+knows the maximum amount of data available to be requested by the
+parser and the actual amount of data which has already been
+requested by the parser.
+
+When a file is first opened for parsing by a parser generated by KS,
+a root stream is created. This root stream can be accessed via
+`_root._io` at any time and in any place. In this scenario, `_root`
+returns the top level object defined in a script, and `_io` is a
+method which can be called on an object to return the associated
+stream. The root stream will know the maximum amount of data available
+to be requested by the parser as the file size of the input file which
+is being parsed. Initially, the root stream will know that 0 bits of
+data have been requested by the parser.
```

Comment:
That's a great point, probably one worth adding to the pitfalls section (or troubleshooting or similar) for the few people who may encounter the issue and not understand what is going on.